// @eagleoutice/flowr/cli/repl/server/messages/message-query.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.responseQueryMessage = exports.requestQueryMessage = void 0;
const joi_1 = __importDefault(require("joi"));
const query_1 = require("../../../../queries/query");
exports.requestQueryMessage = {
    type: 'request-query',
    schema: joi_1.default.object({
        type: joi_1.default.string().valid('request-query').required().description('The type of the message.'),
        id: joi_1.default.string().optional().description('If you give the id, the response will be sent to the client with the same id.'),
        filetoken: joi_1.default.string().required().description('The filetoken of the file/data retrieved from the analysis request.'),
        query: (0, query_1.QueriesSchema)().required().description('The query to run on the file analysis information.')
    }).description('Request a query to be run on the file analysis information.')
};
exports.responseQueryMessage = {
    type: 'response-query',
    schema: joi_1.default.object({
        type: joi_1.default.string().valid('response-query').required(),
        id: joi_1.default.string().optional().description('The id of the message, will be the same for the request.'),
        results: joi_1.default.object().required().description('The results of the query.')
    }).description('The response to a query request.')
};
