// @eagleoutice/flowr/cli/repl/server/messages/message-lineage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.responseLineageMessage = exports.requestLineageMessage = void 0;
const joi_1 = __importDefault(require("joi"));
exports.requestLineageMessage = {
    type: 'request-lineage',
    schema: joi_1.default.object({
        type: joi_1.default.string().valid('request-lineage').required().description('The type of the message.'),
        id: joi_1.default.string().optional().description('If you give the id, the response will be sent to the client with the same id.'),
        filetoken: joi_1.default.string().required().description('The filetoken of the file/data retrieved from the analysis request.'),
        criterion: joi_1.default.string().required().description('The criterion to start the lineage from.')
    })
};
exports.responseLineageMessage = {
    type: 'response-lineage',
    schema: joi_1.default.object({
        type: joi_1.default.string().valid('response-lineage').required(),
        id: joi_1.default.string().optional().description('The id of the message, will be the same for the request.'),
        lineage: joi_1.default.array().items(joi_1.default.string()).required().description('The lineage of the given criterion.')
    })
};
