// @eagleoutice/flowr/cli/repl/server/messages/message-hello.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.helloMessageDefinition = void 0;
const joi_1 = __importDefault(require("joi"));
exports.helloMessageDefinition = {
    type: 'hello',
    schema: joi_1.default.object({
        type: joi_1.default.string().required().valid('hello').description('The type of the hello message.'),
        id: joi_1.default.any().forbidden().description('The id of the message is always undefined (as it is the initial message and not requested).'),
        clientName: joi_1.default.string().required().description('A unique name that is assigned to each client. It has no semantic meaning and is only used/useful for debugging.'),
        versions: joi_1.default.object({
            flowr: joi_1.default.string().required().description('The version of the flowr server running in semver format.'),
            r: joi_1.default.string().required().description('The version of the underlying R shell running in semver format.'),
            engine: joi_1.default.string().required().description('The parser backend that is used to parse the R code.')
        }).required()
    }).required()
};
