// @eagleoutice/flowr/cli/repl/server/messages/message-slice.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.responseSliceMessage = exports.requestSliceMessage = void 0;
const Joi = __importStar(require("joi"));
exports.requestSliceMessage = {
    type: 'request-slice',
    schema: Joi.object({
        type: Joi.string().valid('request-slice').required().description('The type of the message.'),
        id: Joi.string().optional().description('The id of the message, if you passed one in the request.'),
        filetoken: Joi.string().required().description('The filetoken of the file to slice must be the same as with the analysis request.'),
        criterion: Joi.array().items(Joi.string()).min(0).required().required().description('The slicing criteria to use.'),
    })
};
exports.responseSliceMessage = {
    type: 'response-slice',
    schema: Joi.object({
        type: Joi.string().valid('response-slice').required().description('The type of the message.'),
        id: Joi.string().optional().description('The id of the message, if you passed one in the request.'),
        results: Joi.object().required().description('The results of the slice (one field per step slicing step).')
    }).description('The response to a slice request.')
};
