// @eagleoutice/flowr/cli/repl/server/messages/message-repl.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.responseExecuteReplEndMessage = exports.responseExecuteReplIntermediateMessage = exports.requestExecuteReplExpressionMessage = void 0;
const Joi = __importStar(require("joi"));
exports.requestExecuteReplExpressionMessage = {
    type: 'request-repl-execution',
    schema: Joi.object({
        type: Joi.string().valid('request-repl-execution').required().description('The type of the message.'),
        id: Joi.string().optional().description('The id of the message, will be the same for the request.'),
        ansi: Joi.boolean().optional().description('Should ansi formatting be enabled for the response? Is `false` by default.'),
        expression: Joi.string().required().description('The expression to execute.'),
    })
};
exports.responseExecuteReplIntermediateMessage = {
    type: 'response-repl-execution',
    schema: Joi.object({
        type: Joi.string().valid('response-repl-execution').required().description('The type of the message.'),
        id: Joi.string().optional().description('The id of the message, will be the same for the request.'),
        stream: Joi.string().valid('stdout', 'stderr').required().description('The stream the message is from.'),
        result: Joi.string().required().description('The output of the execution.')
    })
};
exports.responseExecuteReplEndMessage = {
    type: 'end-repl-execution',
    schema: Joi.object({
        type: Joi.string().valid('end-repl-execution').required().description('The type of the message.'),
        id: Joi.string().optional().description('The id of the message, will be the same for the request.')
    })
};
