// @eagleoutice/flowr/documentation/doc-util/doc-files.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowrVsCode = exports.FlowrCodecovRef = exports.FlowrDockerRef = exports.FlowrNpmRef = exports.FlowrWikiBaseRef = exports.RemoteFlowrFilePathBaseRef = exports.FlowrSiteBaseRef = exports.FlowrGithubBaseRef = void 0;
exports.getFilePathMd = getFilePathMd;
exports.getFileContentFromRoot = getFileContentFromRoot;
exports.linkFlowRSourceFile = linkFlowRSourceFile;
const fs_1 = __importDefault(require("fs"));
exports.FlowrGithubBaseRef = 'https://github.com/flowr-analysis';
exports.FlowrSiteBaseRef = 'https://flowr-analysis.github.io/flowr';
exports.RemoteFlowrFilePathBaseRef = `${exports.FlowrGithubBaseRef}/flowr/tree/main/`;
exports.FlowrWikiBaseRef = `${exports.FlowrGithubBaseRef}/flowr/wiki`;
exports.FlowrNpmRef = 'https://www.npmjs.com/package/@eagleoutice/flowr';
exports.FlowrDockerRef = 'https://hub.docker.com/r/eagleoutice/flowr';
exports.FlowrCodecovRef = 'https://app.codecov.io/gh/flowr-analysis/flowr';
exports.FlowrVsCode = 'https://marketplace.visualstudio.com/items?itemName=code-inspect.vscode-flowr';
function getFilePathMd(path) {
    // we go one up as we are in doc-util now :D #convenience
    const fullpath = require.resolve('../' + path);
    // normalize path separators so that this is consistent when testing on windows
    const cwd = process.cwd().replaceAll('\\', '/');
    const relative = fullpath.replaceAll('\\', '/').replace(cwd, '.');
    /* remove project prefix */
    return `[\`${relative}\`](${exports.RemoteFlowrFilePathBaseRef}${relative})`;
}
function getFileContentFromRoot(path) {
    const fullpath = require.resolve('../../../' + path);
    return fs_1.default.readFileSync(fullpath, 'utf-8');
}
function linkFlowRSourceFile(path) {
    return `[${path}](${exports.RemoteFlowrFilePathBaseRef}/${path})`;
}
