// @eagleoutice/flowr/core/steps/all/core/11-normalize-tree-sitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NORMALIZE_TREE_SITTER = void 0;
const print_1 = require("../../../print/print");
const normalize_printer_1 = require("../../../print/normalize-printer");
const parser_1 = require("../../../../r-bridge/lang-4.x/ast/parser/json/parser");
const _10_normalize_1 = require("./10-normalize");
function processor(results, input) {
    return (0, parser_1.normalizeTreeSitter)(results['parse'], input.getId, (0, _10_normalize_1.getCurrentRequestFile)(input.request));
}
exports.NORMALIZE_TREE_SITTER = {
    name: 'normalize',
    humanReadableName: 'normalize tree-sitter tree',
    description: 'Normalize the AST to flowR\'s AST',
    processor,
    executed: 0 /* PipelineStepStage.OncePerFile */,
    printer: {
        [0 /* StepOutputFormat.Internal */]: print_1.internalPrinter,
        [2 /* StepOutputFormat.Json */]: normalize_printer_1.normalizedAstToJson,
        [5 /* StepOutputFormat.RdfQuads */]: normalize_printer_1.normalizedAstToQuads,
        [3 /* StepOutputFormat.Mermaid */]: normalize_printer_1.printNormalizedAstToMermaid,
        [4 /* StepOutputFormat.MermaidUrl */]: normalize_printer_1.printNormalizedAstToMermaidUrl
    },
    dependencies: ['parse'],
    requiredInput: undefined
};
