// @eagleoutice/flowr/cli/repl/commands/repl-normalize.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeStarCommand = exports.normalizeCommand = void 0;
const default_pipelines_1 = require("../../../core/steps/pipeline/default-pipelines");
const retriever_1 = require("../../../r-bridge/retriever");
const ast_1 = require("../../../util/mermaid/ast");
const clipboard = Promise.resolve().then(() => __importStar(require('clipboardy')));
const ansi_1 = require("../../../util/ansi");
async function normalize(parser, remainingLine) {
    return await (0, default_pipelines_1.createNormalizePipeline)(parser, {
        request: (0, retriever_1.requestFromInput)(remainingLine.trim())
    }).allRemainingSteps();
}
function handleString(code) {
    return code.startsWith('"') ? JSON.parse(code) : code;
}
function formatInfo(out, type, timing) {
    return out.formatter.format(`Copied ${type} to clipboard (normalize: ${timing}ms).`, { color: 7 /* Colors.White */, effect: ansi_1.ColorEffect.Foreground, style: 3 /* FontStyles.Italic */ });
}
exports.normalizeCommand = {
    description: `Get mermaid code for the normalized AST of R code, start with '${retriever_1.fileProtocol}' to indicate a file`,
    usageExample: ':normalize',
    aliases: ['n'],
    script: false,
    fn: async (output, shell, remainingLine) => {
        const result = await normalize(shell, handleString(remainingLine));
        const mermaid = (0, ast_1.normalizedAstToMermaid)(result.normalize.ast);
        output.stdout(mermaid);
        try {
            (await clipboard).default.writeSync(mermaid);
            output.stdout(formatInfo(output, 'mermaid url', result.normalize['.meta'].timing));
        }
        catch { /* do nothing this is a service thing */ }
    }
};
exports.normalizeStarCommand = {
    description: 'Returns the URL to mermaid.live',
    usageExample: ':normalize*',
    aliases: ['n*'],
    script: false,
    fn: async (output, shell, remainingLine) => {
        const result = await normalize(shell, handleString(remainingLine));
        const mermaid = (0, ast_1.normalizedAstToMermaidUrl)(result.normalize.ast);
        output.stdout(mermaid);
        try {
            (await clipboard).default.writeSync(mermaid);
            output.stdout(formatInfo(output, 'mermaid url', result.normalize['.meta'].timing));
        }
        catch { /* do nothing this is a service thing */ }
    }
};
